### ISSUE TITLE
`mgm_multi` bug

### ISSUE THREAD

**Issue by glostis** [cmla/s2p#20]
I was running `s2p` on branch `dev` and encountered the following error when running `s2p/bin/mgm_multi`:
```
Assertion failed: (min<max), function init, file ./dvec.cc, line 57.
```

I first tried to run the same command on the same images using an `mgm_multi` binary compiled directly from https://github.com/gfacciol/mgm branch `multiscale` (by simply running `make` in `mgm`). With this binary, the command succeeds.

Since `s2p/3rdparty/mgm_multi` (branch `dev`) is at the same commit as `mgm` (branch `multiscale`), the only reason that could explain this bug lies in the `makefile` of `s2p`.

After some investigation, I have found the offending commit to be: https://github.com/cmla/s2p/commit/623321c8bba8a505bc259a6946901d79a839f75b#diff-cd5dbf3629c558b0104ba8f0937d6816L57

In this commit, the way `mgm_multi` is compiled is completely changed. Before this commit, the `mgm_multi` binary works fine, and at this commit, `mgm_multi` crashes with the error above.

I am attaching a zip ([mgm_bug.zip](https://github.com/cmla/s2p/files/3427327/mgm_bug.zip)) showing an example of this behavior.
Run `install.sh` to clone `mgm` at branch `multiscale` and compile 3 different `mgm_multi` binaries:
- one compiled with the `s2p` `makefile` at the offending commit (`623321c8bba8a505bc259a6946901d79a839f75b_mgm_multi`)
- one compiled with the `s2p` `makefile` just before the offending commit (`fb9ed0f73cf655175c8736282d00f99730974193_mgm_multi`)
- one compiled using `mgm`'s `makefile` directly (`original_mgm_multi`)

Run `test.sh` to see how the different binaries behave on the same command:
- the first binary fails
- the two other binaries do not fail

I do not know much about compilation so I am not able to understand the reason of this crash, but I would like to have your help on this.

**Comment by glostis** [comment]
Response from @mnhrdt:

> This is because the -DNDEBUG option is not enabled by default on the
> s2p makefile, and when running the code the assertion fails.  The
> NDEBUG macro is used to disable assertions on C and C++ code.
> 
> The problem can be solved by adding -DNDEBUG on the CXXFLAGS variable
> on the s2p makefile.  However, this just hides the problem, does not
> really solve it.
> 
> I guess the assertion on dvec.cc is wrong, and it should be
> "assert(dmin<=dmax)" to account for the possibility of an empty range.

Replacing the assert with `assert(min<=max)` does indeed solve the problem, thanks @mnhrdt.
@gfacciol if you agree, could we push this fix to `mgm_multi`, and update the submodule in `s2p`?

**Comment by gfacciol** [comment]
Yes, I will push it on mgm. But this assert shows that there is a problem
somewhere else. Indeed, the diaparity ranges should always be larger than
1.
The ranges computed internally by mgm are always dilated by a consant > 1.
It might be interesting to map out  the pixels where this assertion fails.

You're feeding the Dmin Dmaz values directly to the program?

Best
Gabriele

**Comment by carlodef** [comment]
Fixed by 0896685.
